File: Aigamo.Petunia/JavaScriptImport.cs
namespace Aigamo.Petunia;

internal abstract class JavaScriptImport
{
	public string ModuleName { get; }

	protected JavaScriptImport(string moduleName)
	{
		ModuleName = moduleName;
	}

	public sealed class JavaScriptNamedImport : JavaScriptImport
	{
		private readonly List<string> _exports = new();

		public JavaScriptNamedImport(string moduleName)
			: base(moduleName)
		{
		}

		public JavaScriptNamedImport AddExport(string? export)
		{
			if (export is not null) _exports.Add(export);
			return this;
		}

		public JavaScriptNamedImport AddExports(params string?[]? exports)
		{
			if (exports is not null) _exports.AddRange(exports.WhereNotNull());
			return this;
		}

		public override string ToString()
		{
			var exports = string.Join(", ", _exports.Distinct().Order());
			return $"import {{ {exports} }} from '{ModuleName}';";
		}
	}

	public sealed class JavaScriptDefaultImport : JavaScriptImport
	{
		public string DefaultExport { get; }

		public JavaScriptDefaultImport(string moduleName, string defaultExport)
			: base(moduleName)
		{
			DefaultExport = defaultExport;
		}

		public override string ToString()
		{
			return $"import {DefaultExport} from '{ModuleName}';";
		}
	}
}

internal sealed class JavaScriptImports
{
	private readonly List<JavaScriptImport> _imports = new();

	public JavaScriptImports AddImport(JavaScriptImport? import)
	{
		if (import is not null) _imports.Add(import);
		return this;
	}

	public JavaScriptImports AddImports(params JavaScriptImport?[]? imports)
	{
		if (imports is not null) _imports.AddRange(imports.WhereNotNull());
		return this;
	}

	public override string ToString()
	{
		return string.Join("\n", _imports.OrderBy(import => import.ModuleName));
	}
}

File: src/core/JavaScriptImport.ts
import { orderBy, uniq } from 'lodash-es';

abstract class JavaScriptImport {
	constructor(readonly moduleName: string) {}

	abstract toString(): string;
}

class JavaScriptNamedImport extends JavaScriptImport {
	private readonly names: string[] = [];

	addNamedExport = (name?: string): this => {
		if (name !== undefined) this.names.push(name);
		return this;
	};

	addNamedExports = (...names: (string | undefined)[]): this => {
		if (names !== undefined) {
			this.names.push(
				...names.filter((name): name is string => name !== undefined),
			);
		}
		return this;
	};

	toString = (): string => {
		const namesString = orderBy(uniq(this.names), (name) => name).join(
			', ',
		);
		return `import { ${namesString} } from '${this.moduleName}';`;
	};
}

class JavaScriptDefaultImport extends JavaScriptImport {
	constructor(
		readonly defaultExport: string,
		moduleName: string,
		readonly assertions = '',
	) {
		super(moduleName);
	}

	toString = (): string => {
		return `import ${this.defaultExport} from '${this.moduleName}'${this.assertions};`;
	};
}

interface FormatOptions {
	newLine: string;
}

export class JavaScriptImports {
	private readonly values: JavaScriptImport[] = [];

	addImport = (value?: JavaScriptImport): this => {
		if (value !== undefined) this.values.push(value);
		return this;
	};

	addDefaultImport = (
		defaultExport: string,
		moduleName: string,
		assertions?: string,
	): this => {
		return this.addImport(
			new JavaScriptDefaultImport(defaultExport, moduleName, assertions),
		);
	};

	addNamedImport = (
		moduleName: string,
		callback: (builder: JavaScriptNamedImport) => void,
	): this => {
		const builder = new JavaScriptNamedImport(moduleName);
		callback(builder);
		return this.addImport(builder);
	};

	addImports = (...values: (JavaScriptImport | undefined)[]): this => {
		if (values !== undefined) {
			this.values.push(
				...values.filter(
					(value): value is JavaScriptImport => value !== undefined,
				),
			);
		}
		return this;
	};

	toFormattedString = ({ newLine }: FormatOptions): string => {
		return orderBy(this.values, (value) => value.moduleName).join(newLine);
	};
}
